x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleTask.java

```diff
@@ -78,4 +94,80 @@ public void onFailure(Exception e) {
                .filter(policyMeta -> jobId.equals(SnapshotLifecycleService.getJobId(policyMeta)))
                .findFirst());
     }
+
+    private static class WriteJobStatus extends ClusterStateUpdateTask {
+        private static final ToXContent.Params STACKTRACE_PARAMS =
+            new ToXContent.MapParams(Collections.singletonMap(REST_EXCEPTION_SKIP_STACK_TRACE, "false"));
+
+        private final String policyName;
+        private final long timestamp;
+        private final Optional<Exception> exception;
+
+        private WriteJobStatus(String policyName, long timestamp, Optional<Exception> exception) {
+            this.policyName = policyName;
+            this.exception = exception;
+            this.timestamp = timestamp;
+        }
+
+        static WriteJobStatus success(String policyId, long timestamp) {
+            return new WriteJobStatus(policyId, timestamp, Optional.empty());
+        }
+
+        static WriteJobStatus failure(String policyId, long timestamp, Exception exception) {
+            return new WriteJobStatus(policyId, timestamp, Optional.of(exception));
+        }
+
+        private String exceptionToString() throws IOException {
+            if (exception.isPresent()) {
```

Comment:
Only partially true - I'm changing this to use try-with-resources as in your example, but it can't be done with `Optional.map()` because that doesn't play nice with things that can throw exceptions.

Also, I copied this a bit from ILM and I think we should do this here, too:
https://github.com/elastic/elasticsearch/blob/647e2256984abd44d5d0820bc742a0af2061a2b8/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java#L334-L343